proposals/0228-fix-expressiblebystringinterpolation.md

```diff
@@ -88,6 +88,55 @@ String interpolation is the only remaining client of this type-checker entry poi
 
 </details>
 
+### Potential uses
+
+An improved string interpolation design could enable future standard library and Apple framework features; here are a few ideas. Note that we are not proposing these now and might not ever propose them in the future—they're just examples.
+
+#### Constructing formatted strings
+
+There are a number of approaches we could take to formatting values interpolated into strings. Here are a few examples with numbers:
+
+```swift
+// Use printf-style format strings:
+"The price is $\(cost, format: "%.2f")"
+
+// Use UTS #35 number formats:
+"The price is \(cost, format: "¤###,##0.00")"
+
+// Use Foundation.NumberFormatter, or a new type-safe native formatter:
+"The price is \(cost, format: moneyFormatter)"
+
+// Mimic String.init(_:radix:uppercase:)
+"The checksum is 0x\(checksum, radix: 16)"
+```
+
+You could imagine analogous formatting tools for other types, like `Data`, `Date`, or even just `String` itself.
+
+#### Logging
+
+Some logging facilities restrict the kinds of data that can be logged or require extra metadata on certain values; a more powerful interpolation feature could support that:
+
+```swift
+log("Processing \(public: tagName) tag containing \(private: contents)")
+```
+
+#### Constructing attributed strings
+
+`NSAttributedString` or a value-type wrapper around it could allow users to interpolate dictionaries of attributes to enable and disable them:
+
+```swift
+"\([.link: supportURL])Click here\([.link: nil]) to visit our support site"
+```
+
+#### Localization
+
+A `LocalizableString` type could be expressed by a string literal, which would be used to generate a format string key and a list of arguments; converting a `LocalizableString` to an ordinary `String` would look up the key in a `Bundle`'s localization table, then format the value with the arguments.
+
+```swift
+alert.messageText = Bundle.main.localized("The document “\(name)” could not be saved.")
+// Looks up "The document “%@” could not be saved."
```

Comment:
Would a better formulation be something akin to:

```swift
let localizedMessage: LocalizableString = “The document \(key: name, bundle: Bundle.main) could not be saved"
alert.messageText = String(localizedMessage)
```

Reply:
When I've written things like this in the past, I always used a helper function with a signature `(LocalizableString) -> String` so you wouldn't need to write types explicitly. But writing it the way you describe might make help explain how this would work.

Reply:
Oh, I see now that you're suggesting the interpolation would be different.

I don't think that interpolation format really matches how Foundation localization works—what we're doing is building a printf-style format string from the literal segments and types of interpolations, then looking that string up in a localization table to get its equivalent in the user's language, and finally using the localized format string to format the arguments. `name` is not something being looked up; it's something being substituted in.

But I'm thinking I'll replace the sample with a clearer equivalent:

```swift
// Builds a LocalizableString(key: "The document “%@” could not be saved.", arguments: [name])
let message: LocalizableString = "The document “\(name)” could not be saved."
alert.messageText = String(localized: message)
```

This `String(localized:)` is a nearly drop-in replacement for `NSLocalizedString(_:comment:)` which supports interpolation, so I think working Cocoa programmers will understand the appeal.